core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -410,21 +422,78 @@ public void index(Index index) {
+             * to detect if it has potentially been added before. We use the documents timestamp for this since it's something
+             * that:
+             *  - doesn't change per document
+             *  - is preserved in the transaction log
+             *  - and is assigned before we start to index / replicate
+             * NOTE: it's not important for this timestamp to be consistent across nodes etc. it's just a number that is in the common
+             * case increasing and can be used in the failure case when we retry and resent documents to establish a happens before relationship.
+             * for instance:
+             *  - doc A has timestamp 10, autoGeneratedID = true, canHaveDuplicates = true
+             *  - doc B has timestamp 9, autoGeneratedID = true, canHaveDuplicates = true
+             *
+             *  while both docs are in in flight, we disconnect on one node, reconnect and send doc A again
+             *  - now doc A' has timestamp 10, autoGeneratedID = true, canHaveDuplicates = true
+             *
+             *  if A' arrives on the shard first we update highestDeOptimizeAddDocumentTimestamp to 10 and use update document. All subsequent
+             *  documents that arrive (A and B) will also use updateDocument since their timestamps are less than highestDeOptimizeAddDocumentTimestamp.
+             *  While this is not strictly needed for doc B it is just much simpler to implement since it will just de-optimize some doc in the worst case.
+             *
+             *  if A arrives on the shard first we use addDocument since highestDeOptimizeAddDocumentTimestamp is < 10. A` will then just be skipped or calls
+             *  updateDocument.
+             */
+            final boolean skipVersionCheck;
+            if (index.isAutoGeneratedID()
+                && index.version() == Versions.MATCH_ANY
+                && index.versionType() == VersionType.INTERNAL) {
+                // this optimization is only enabled for version = -1 && type=internal and nothing else
+                long deOptimizeTimestamp = highestDeOptimizeAddDocumentTimestamp.get();
+                if (index.canHaveDuplicates()) {
+                    useAddDocument = false;
+                    do {
+                        deOptimizeTimestamp = highestDeOptimizeAddDocumentTimestamp.get();
+                        if (deOptimizeTimestamp >= index.timestamp()) {
+                            break;
+                        }
+                    } while(highestDeOptimizeAddDocumentTimestamp.compareAndSet(deOptimizeTimestamp, index.timestamp()) == false);
+                    assert highestDeOptimizeAddDocumentTimestamp.get() >= index.timestamp();
+                } else {
+                    useAddDocument = deOptimizeTimestamp < index.timestamp();
+                }
+                skipVersionCheck = useAddDocument == false; // we skip the version check if we can't use the optimization and just
```

Comment:
I'm not sure this is correct. If the optimization is turned off because it's unsafe (i.e., useAddDocument is). We can do two things:

1) check ourselves by loading versions from the index or memory map and see if we already have the document. If we do, with don't re-index but respond with "OK".
2) don't check the current versions but rather index a new document (with version set to 1 etc) to lucene using updateDocument.

I'm in favor of two, allowing use to nuke the version map.

For clarity, right now the code does the normal thing which means loading versions and either reindexing with a version = 2 on the primary or rejecting the indexing operation on the replica.